### ISSUE TITLE
Extract args are unintuitive

### ISSUE THREAD

**Issue by twizmwazin** [angr/claripy#502]
### Description

The args to extract aren't very intuitive: (high (inclusive), low (inclusive), base). This seems like we copied it straight from z3, which did the same thing: https://z3prover.github.io/api/html/namespacez3py.html#a40e9429ef16134a6d9914ecdc2182e8c

What's even better though is that z3 later just overloaded this function for Strings and Sequences, but when they did that they must have been in a clearer state of mind, as for that variant the args are completely different: (base, index, length). I think we should copy those args for BVs too.

I guess this is more intuitive if you are reading the little-endian form of the BV, ex `Extract(7, 0, BVV(0xabcd. 16)) == BVV(0xcd, 8)`. Though even in this case it doesn't feel very consistent, and the logic in get_bytes reflects this.

**Comment by zardus** [comment]
It's definitely a hold-over from Z3, but I would advise against changing
Extract, as Z3 _is_ the "industry standard" and that conceptual
compatibility is important. Instead, what if we create a synonym to
Extract, and have it make more sense?

- Yan

**Comment by ltfish** [comment]
Let's not change it as it's going to impact too much existing code, much of which is not written by us. The ship has sailed at this point.

**Comment by twizmwazin** [comment]
I disagree with the general idea that it is too late to change something, holding on to tech debt because it is old leads to bad software that is hard to learn, hard to maintain, makes usage generally unpleasant and eventually kills a project when something better comes along.

But I think this is easy enough to type check at runtime that we could maintain compatibility for a while and give people advanced notice. I could change Extract to either support (bv, offset, size) or the current args, issuing a deprecation warning for the current args. And then after a year or when the rust implementation is ready, we drop the compatibility then, ideally marked with a major version release and a migration guide.

**Comment by twizmwazin** [comment]
Alright yeah if Z3 is also consistent with SMT-LIB2 I think I'm in agreement with you. I'll see about adding a second set of args and maintaining both.

**Comment by twizmwazin** [comment]
Closing this as we will maintain the existing interface for consistency
